pkg/async/cond.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: Cond.go provides a context respecting sync condition
+
+package async
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+// Cond is a sync.Cond that respects context cancellation.
+// It provides equivalent functionality to sync.Cond (excepting there is no `Signal` method), except that
+// the Wait method exits with error if the context cancels.
+type Cond struct {
+	pointer atomic.Pointer[chan struct{}]
+}
+
+// ch returns the channel that Waiters are waiting on, possibly creating one if it doesn't exist.
+func (c *Cond) ch() chan struct{} {
+	t := make(chan struct{})
+	c.pointer.CompareAndSwap(nil, &t)
+	return *c.pointer.Load()
+}
+
+// Wait waits for the state change Broadcast until context ends.
+func (c *Cond) Wait(ctx context.Context) error {
+	select {
+	case <-c.ch():
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Broadcast signals the state change to all Waiters
+func (c *Cond) Broadcast() {
+	// now that we retrieved the channel, new calls to Wait should get a new channel
+	ch := c.pointer.Swap(nil)
+	if ch != nil {
+		close(*ch)
+	}
+}
+
+// WaitForCondition checks if the condition is true or the context is done, otherwise
+// it waits for the state change Broadcast.
+//
+// if it returns without error, it also locks the provided locker and the caller must call the returned function
```

Comment:
Clearly document that the lock is held while running `condition`, and released on error?

Reply:
done